### PR TITLE
Change output of rewards snapshot script

### PR DESCRIPTION
**Pull request by sevanspowell**
[ADP-1382](https://input-output.atlassian.net/browse/ADP-1382)

- Output `voting_public_key`, `stake_pub`, `reward_address`, and `voting_power`, instead of just the Jormungandr address (derived from `voting_public_key`) and `voting_power`.
- Subsume Rewards and Genesis scripts into single Snapshot script.
- No longer reject registrations with value below a given threshold.
- Provide helpful functions for debugging output of the voting-tool.

### CODE REVIEW

**Comment by sevanspowell** [comment]
https://hydra.iohk.io/jobset/Cardano/voting-tools#tabs-evaluations

It seems the evaluation has been running for 2+hrs :/ I'm not sure what's going on here.

Also I got this error in Hydra:

```
in job ‘native.voterRegistrationTarball.x86_64-darwin’:
error: a 'x86_64-darwin' with features {} is required to build '/nix/store/idzxyipy2nzcij2k9nxzzj3dn6gar97c-musl-x86_64-unknown-linux-musl-1.2.2.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}
```

Hence I've added this commit to workaround it: https://github.com/input-output-hk/voting-tools/pull/32/commits/50161c150943fdb6bcafe4608f35cab14d6f78a3

File: src/Snapshot.hs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+getBlockZeroDate :: IO UTCTime
+getBlockZeroDate = do
+  (UTCTime day dayTime) <- getCurrentTime
+  let
+    (TimeOfDay hr _ _ ) = timeToTimeOfDay dayTime
+    blockZeroDate       = UTCTime day (timeOfDayToTime $ TimeOfDay hr 0 0)
+  pure blockZeroDate
+
+setInitialFunds :: [Fund] -> Aeson.Value -> Aeson.Value
+setInitialFunds funds = _Object %~ HM.insert "initial" (Aeson.toJSON funds)
```
Comment (rvl) [inline src/Snapshot.hs:31]:
```suggestion
setInitialFunds funds = _Object . at "initial" ?~ toJSON funds
```

File: src/Snapshot.hs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+setBlockZeroDate :: UTCTime -> Aeson.Value -> Aeson.Value
+setBlockZeroDate time =
+  (key "blockchain_configuration"
+    %~ (key "block0_date"
+      .~ (Aeson.Number
+         . (fromIntegral :: Integer -> Scientific)
```
Comment (rvl) [inline src/Snapshot.hs:38]:
```suggestion
      .~ (toJSON
```
(I think this would work)

File: src/Snapshot.hs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+unsafeDecodeJSONFile :: FilePath -> IO (Aeson.Value)
+unsafeDecodeJSONFile fp = do
+  result <- Aeson.eitherDecodeFileStrict' fp
+  case result of
+    Left err                               -> error err
+    Right (snapshotTemplate :: Aeson.Value) -> pure snapshotTemplate
+
+getAllFundContents :: Aeson.Value -> Aeson.Value
+getAllFundContents x =
+  x & (^.. key "initial" . values . key "fund". _Array)
```
Comment (rvl) [inline src/Snapshot.hs:52]: I get rather confused by chains of `&`.
It's hard to understand what this function does.

File: src/Snapshot.hs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    & mconcat
+    & Vector.toList
+    & sortOn (^?! _Object . at "address")
+    & Vector.fromList
+    & Aeson.Array
+
+getSnapshotContents :: NetworkId -> Aeson.Value -> Aeson.Value
+getSnapshotContents networkId x =
+  x & values . _Object %~ sans "stake_public_key" . sans "reward_address"
+    & values . key "voting_public_key" . _JSON %~ (\(votePub :: VotingKeyPublic) -> addressFromVotingKeyPublic networkId votePub)
```
Comment (rvl) [inline src/Snapshot.hs:62]:
```suggestion
    & values . key "voting_public_key" . _JSON %~ addressFromVotingKeyPublic networkId
```

File: src/Snapshot.hs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                   votePower = m ^?! at "voting_power" . _Just . _Number
+                 in
+                   HM.insertWith (+) votePub votePower acc
+              ) mempty
+    & HM.toList
+    & sortOn fst
+    & foldMap (\(votePub, votePower) -> [ Aeson.Object $ HM.fromList
+        [ ("address", Aeson.String votePub)
+        , ("value", Aeson.Number votePower)]
+        ]
```
Comment (rvl) [inline src/Snapshot.hs:76]:
```suggestion
    & map (\(votePub, votePower) -> Aeson.object [ "address" .= votePub, "value" .= votePower ]
```

File: test/Test/Cardano/API/Jormungandr.hs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+tests :: TestTree
+tests = testGroup "Test.Cardano.API.Jormungandr"
+  [ testProperty "VotingKeyPublic/Jormungandr.Address roundtrip" prop_votingKeyPublic_jormungandrAddress_roundtrip
+  ]
+
+prop_votingKeyPublic_jormungandrAddress_roundtrip :: Property
+prop_votingKeyPublic_jormungandrAddress_roundtrip = property $ do
+  votepub <- forAll Gen.votingKeyPublic
+  tripping votepub (addressFromVotingKeyPublic Mainnet) (votingKeyPublicFromAddress)
```
Comment (rvl) [inline test/Test/Cardano/API/Jormungandr.hs:22]: Nice

File: test/Test/Snapshot.hs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.HUnit (Assertion, testCase, (@?=))
+
+import           Snapshot
+
+import qualified Data.Aeson as Aeson
+import qualified Data.HashMap.Strict as HM
+
+tests :: TestTree
+tests = testGroup "Test.Snapshot"
```
Comment (rvl) [inline test/Test/Snapshot.hs:25]: Also looks good. :+1:

**Review by rvl — APPROVED** [review]
LGTM, thanks @sevanspowell.

**Comment by disassembler** [comment]
bors ping

**Pull request merged**